src/ic-ledger-types/src/lib.rs

```diff
@@ -0,0 +1,303 @@
+    InsufficientFunds { balance: Tokens },
+    TxTooOld { allowed_window_nanos: u64 },
+    TxCreatedInFuture,
+    TxDuplicate { duplicate_of: BlockIndex },
+}
+
+impl fmt::Display for TransferError {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        match self {
+            Self::BadFee { expected_fee } => {
+                write!(f, "transaction fee should be {}", expected_fee)
+            }
+            Self::InsufficientFunds { balance } => {
+                write!(
+                    f,
+                    "the debit account doesn't have enough funds to complete the transaction, current balance: {}",
+                    balance
+                )
+            }
+            Self::TxTooOld {
+                allowed_window_nanos,
+            } => write!(
+                f,
+                "transaction is older than {} seconds",
+                allowed_window_nanos / 1_000_000_000
+            ),
+            Self::TxCreatedInFuture => write!(f, "transaction's created_at_time is in future"),
+            Self::TxDuplicate { duplicate_of } => write!(
+                f,
+                "transaction is a duplicate of another transaction in block {}",
+                duplicate_of
+            ),
+        }
+    }
+}
+
+/// Calls the "account_balance" method on the specified canister.
+///
+/// # Example
+/// ```no_run
+/// use ic_cdk::api::{caller, call::call};
+/// use ic_ledger_types::{AccountIdentifier, AccountBalanceArgs, Tokens, DEFAULT_SUBACCOUNT, MAINNET_LEDGER_CANISTER_ID, account_balance};
+///
+/// async fn check_callers_balance() -> Tokens {
+///   account_balance(
+///     MAINNET_LEDGER_CANISTER_ID,
+///     AccountBalanceArgs {
+///       account: AccountIdentifier::new(&caller(), &DEFAULT_SUBACCOUNT)
+///     }
+///   ).await.expect("call to ledger failed")
+/// }
+/// ```
+pub async fn account_balance(
+    ledger_canister_id: Principal,
+    args: AccountBalanceArgs,
+) -> CallResult<Tokens> {
+    let (icp,) = ic_cdk::call(ledger_canister_id, "account_balance", (args,)).await?;
+    Ok(icp)
+}
+
+/// Calls the "transfer" method on the specified canister.
+/// # Example
+/// ```no_run
+/// use ic_cdk::api::{caller, call::call};
+/// use ic_ledger_types::{AccountIdentifier, BlockIndex, Memo, TransferArgs, Tokens, DEFAULT_SUBACCOUNT, DEFAULT_FEE, MAINNET_LEDGER_CANISTER_ID, transfer};
+///
+/// async fn transfer_to_caller() -> BlockIndex {
+///   transfer(
+///     MAINNET_LEDGER_CANISTER_ID,
+///     TransferArgs {
+///       memo: Memo(0),
+///       amount: Tokens::from_e8s(1_000_000),
+///       fee: DEFAULT_FEE,
+///       from_subaccount: None,
+///       to: AccountIdentifier::new(&caller(), &DEFAULT_SUBACCOUNT),
+///       created_at_time: None,
+///     }
+///   ).await.expect("call to ledger failed").expect("transfer failed")
+/// }
+/// ```
+pub async fn transfer(
+    ledger_canister_id: Principal,
+    args: TransferArgs,
+) -> CallResult<TransferResult> {
```

Comment:
We could consider flattening the errors given that `TransferError ` is specific for this method. I don't find particularly nice the double handling, e.g. the double `.expect` above, and I don't think it adds anything useful.

Reply:
I'm not sure adding a new error type like this:
```rust
pub enum TransferCallError {
    CallError(CallError),
    TransferError(TransferError),
}
```
 is worth it. This type is equivalent to the current one, and there is no huge difference in the error handling:
```rust
match transfer(...) {
  Ok(Ok(block_index)) => all_is_fine(),
  Ok(Err(TransferError::...)) => handle_transfer_error(),
  Err(CallError::...) => handle_call_error(),
}
```
That doesn't look particularly bad to me.
I'd prefer to keep the interface as close to the .did file as possible.

Reply:
I meant adding a variant to TransferError so that one can do

```rust
match transfer(....) {
  Ok(block_index) => all_is_fine(),
  Err(TransferError::CallError(...)) => handle_call_error(),
  Err(TransferError::...) => handle_other_errors(),
}
```

Reply:
That seems confusing to me, because this way `TransferError` doesn't match the .did file

Reply:
You are right